### PR TITLE
[NeMo-UX] Turn on mcore performance optimizations

### PR DESCRIPTION
**Pull request by JimmyZhang12**
# What does this PR do ?
Turns on various communication/compute overlap optimizations:

- Pipeline parallel overlap, data parallel overlap is well supported and this PR turns it on by default when available

- Tensor parallel overlap is not officially productized in TE, so its not guaranteed to work on all clusters. So it is disabled by default. It can be turned on by setting the `NEMO_ENABLE_TP_COMM_OVERLAP` environment variable. 
Currently TP overlap only supports dense models, so Ive only added a TP overlap config for LLAMA70B, and can add more as more model recipes are added. 

- Wgrad deferral, which overlaps wgrads over the pipeline bubble, has to be tuned per model, so this defaults false.

I tested the full 70b, 8b performance recipes using NeMo-Run as saw no differences to the loss, but a ~10% improvement in performance.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/recipes/llama3_70b.py [hunk 1]
```diff
@@ -44,6 +45,8 @@ def trainer(
         virtual_pipeline_model_parallel_size=virtual_pipeline_parallelism,
         context_parallel_size=context_parallelism,
         sequence_parallel=sequence_parallelism,
+        tp_comm_overlap=False,
```
Comment (gautham-kollu) [inline nemo/collections/llm/recipes/llama3_70b.py:48]: Can we make bundle both tp_comm_overlap and  
tp_comm_overlap_cfg into one dataclass ?

```
class TPCommOverlap: 
     enabled : Default(False)
     cfg_settings : Optional[]

```

Why : Helped keep all the relevant info together (Encapsulation)

File: nemo/lightning/megatron_parallel.py [hunk 2]
```diff
@@ -475,6 +482,37 @@ def infer_num_microbatches(self, data: Union[DataT, Iterator[DataT], List[Iterat
...
+            transformer_engine.pytorch.module.base.initialize_ub(
+                shape=input_shape,
+                tp_size=tp_size,
+                use_fp8=fp8,
+                ub_cfgs=ub_cfgs,
+            )
+        except Exception as error:
+            raise Exception(f"Tensor parallel overlap: userbuffer initialization failed with {error}")
+
+        self.tensor_parallel_overlap_need_init = False
```
Comment (gautham-kollu) [inline nemo/lightning/megatron_parallel.py:514]: Why is this required ?

File: nemo/lightning/pytorch/strategies.py [hunk 3]
```diff
@@ -208,6 +226,23 @@ def __init__(
...
             raise ValueError(f"Invalid DDP type: {ddp}")
 
+        if self.tp_comm_overlap:
+            if (
+                self.tensor_model_parallel_size < 2
+                or not self.sequence_parallel
+                or not HAVE_TE
+                or not transformer_engine.pytorch.cpp_extensions.userbuf_comm_available()
+            ):
+                logging.warning("Disabling tensor parallel overlap due to incompatible setup.")
```
Comment (gautham-kollu) [inline nemo/lightning/pytorch/strategies.py:236]: Helpful to print the conditional that failed, so users have a direct remedy from the log.

File: nemo/lightning/megatron_parallel.py [hunk 4]
```diff
@@ -475,6 +482,37 @@ def infer_num_microbatches(self, data: Union[DataT, Iterator[DataT], List[Iterat
 
         raise ValueError("Cannot infer `num_microbatches` from data, please specify it manually")
 
+    def init_tensor_parallel_overlap(self, micro_batch_size, sequence_length):
+        import transformer_engine
```
Comment (gautham-kollu) [inline nemo/lightning/megatron_parallel.py:486]: Do we need import guards to check if the TE is available ?

**Review by gautham-kollu — CHANGES_REQUESTED** [review]

File: nemo/lightning/megatron_parallel.py (hunk 2 above)
Comment (JimmyZhang12) [reply nemo/lightning/megatron_parallel.py:514]: For the init we need to microbatch size and sequence length, and in megatron_parallel.py it seems this is only available in [in the forward function](https://github.com/NVIDIA/NeMo/pull/10209/files/95267ec814f617adc00564b10fef622cbbdb8d6b#diff-3a143831f2dddc93960ceef806f3dd2b06b29de548b3ecc0669cf3ab6969e85eR238)

So I have this so that init only occurs once there, maybe there is somewhere init should go so that we woudn't need this?

File: nemo/lightning/megatron_parallel.py [hunk 5]
```diff
@@ -475,6 +482,37 @@ def infer_num_microbatches(self, data: Union[DataT, Iterator[DataT], List[Iterat
...
+            sequence_length * micro_batch_size // cp_size,
+            hidden_size,
+        ]
+
+        try:
+            if ub_cfgs is None:
+                logging.warning(
+                    "Tensor parallel overlap: No overlap config provided. Initializing TP comm overlap with the default config."
+                )
+            transformer_engine.pytorch.module.base.initialize_ub(
```
[the remark this answers is not in the document]
Comment (JimmyZhang12) [reply nemo/lightning/megatron_parallel.py:505]: If no overlap config is provided then it will default to TE's baked in configs, which may not maximize performance, so I have a warning for that here.

File: nemo/collections/llm/recipes/llama3_70b.py [hunk 6]
```diff
@@ -95,6 +96,21 @@ def pretrain_recipe(
     )
 
 
+def pretrain_recipe_performance(
```
Comment (marcromeyn) [inline nemo/collections/llm/recipes/llama3_70b.py:99]: @hemildesai I am wondering if this would better to have as a `run.Config`?

File: nemo/lightning/pytorch/strategies.py [hunk 7]
```diff
@@ -211,6 +240,92 @@ def __init__(
         # used in NVIDIA NGC PyTorch containers
         _strategy_lib.enable_nvidia_optimizations()
 
+    def _apply_overlap_configs(self, comm_overlap_cfg: CommOverlapConfig, cfg):
```
Comment (marcromeyn) [inline nemo/lightning/pytorch/strategies.py:243]: Is there a way to break this out in a Callback or something? Would be great if we can keep things modular

Comment (JimmyZhang12) [reply nemo/lightning/pytorch/strategies.py:243]: These functions have to run before the model, optimizer get initialized so before `setup`, is there a way to do that with  `Callback`?

File: nemo/collections/llm/gpt/model/base.py [hunk 8]
```diff
@@ -113,6 +112,8 @@ class GPTConfig(TransformerConfig, io.IOMixin):
     seq_length: int = 1024
     attention_softmax_in_fp32: bool = False
     masked_softmax_fusion: bool = True
+    cross_entropy_loss_fusion: bool = True
+    tp_comm_overlap_cfg: dict = False
```
Comment (gautham-kollu) [inline nemo/collections/llm/gpt/model/base.py:116]: np: Not a dict

File: nemo/lightning/pytorch/callbacks/megatron_comm_overlap.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+from megatron.core.optimizer import OptimizerConfig
+from pytorch_lightning.callbacks.callback import Callback
+
+from nemo.collections.llm.recipes.tp_overlap_configs.userbuffers import TransformerLayerTPOverlapCfg
+from nemo.lightning.pytorch.strategies.megatron_strategy import MegatronStrategy, ParallelismConfig
+from nemo.utils import logging
+
+HAVE_TE = True
+try:
+    import transformer_engine
```
[the remark this answers is not in the document]
Comment (JimmyZhang12) [reply nemo/lightning/pytorch/callbacks/megatron_comm_overlap.py:15]: Okay since we just need to check the import

File: nemo/lightning/megatron_parallel.py [hunk 10]
```diff
@@ -489,6 +503,41 @@ def infer_num_microbatches(self, data: Union[DataT, Iterator[DataT], List[Iterat
 
         raise ValueError("Cannot infer `num_microbatches` from data, please specify it manually")
 
+    def init_tp_comm_overlap(self, micro_batch_size, sequence_length):
```
Comment (JimmyZhang12) [inline nemo/lightning/megatron_parallel.py:506]: The performance optimizations are all self contained in the `MegatronCommOverlap` plugin now. The only exception is this function which needs to run to initialize some buffers for TP overlap
This function needs to know the microbatch size, but I'm not how we can nicely get that while inside a callback @marcromeyn @gautham-kollu do you have any ideas?

Comment (JimmyZhang12) [reply nemo/lightning/megatron_parallel.py:506]: Nevermind...I learned about `megatron.core.num_microbatches_calculator` so I can probably use that

File: nemo/collections/llm/recipes/llama3_8b.py [hunk 11]
```diff
@@ -95,6 +96,17 @@ def pretrain_recipe(
     )
 
 
+def pretrain_recipe_performance(
```
Comment (marcromeyn) [inline nemo/collections/llm/recipes/llama3_8b.py:99]: Since the factory functions can take in arguments, could we add a bool-field in the llama-8b-pretrain receip to enable comm_overlap which would be False by default?

File: nemo/collections/llm/gpt/model/base.py [hunk 12]
```diff
@@ -113,7 +113,10 @@ class GPTConfig(TransformerConfig, io.IOMixin):
     seq_length: int = 1024
     attention_softmax_in_fp32: bool = False
     masked_softmax_fusion: bool = True
+    cross_entropy_loss_fusion: bool = True
+    gradient_accumulation_fusion: bool = True
     deallocate_pipeline_outputs = True
+    tp_comm_overlap_cfg = None
```
Comment (marcromeyn) [inline nemo/collections/llm/gpt/model/base.py:119]: Do we need this here?

File: nemo/collections/llm/gpt/model/base.py [hunk 13]
```diff
@@ -32,6 +32,14 @@
 
 _, HAVE_TE = safe_import("transformer_engine")
 
+# Gradient accumulation fusion may be enabled if available, for more information see:
+# https://github.com/NVIDIA/Megatron-LM/blob/01945b98d1ea3a2acb5e8301e181a328104f4856/megatron/core/tensor_parallel/layers.py#L575
+_grad_accum_fusion_available = True
+try:
+    import fused_weight_gradient_mlp_cuda
+except ImportError:
+    _grad_accum_fusion_available = False
```
Comment (gautham-kollu) [inline nemo/collections/llm/gpt/model/base.py:41]: np: Consider adding a logline how to install the fused_weight_gradient_mlp_cuda. 

(I see it makes from Apex and is optional)

File: nemo/collections/llm/gpt/model/base.py [hunk 14]
```diff
@@ -32,6 +32,14 @@
 
 _, HAVE_TE = safe_import("transformer_engine")
 
+# Gradient accumulation fusion may be enabled if available, for more information see:
+# https://github.com/NVIDIA/Megatron-LM/blob/01945b98d1ea3a2acb5e8301e181a328104f4856/megatron/core/tensor_parallel/layers.py#L575
+_grad_accum_fusion_available = True
```
Comment (gautham-kollu) [inline nemo/collections/llm/gpt/model/base.py:37]: np: Can we wrap this in a getter api

```
is_grad_accumulation_available()
```

File: nemo/collections/llm/gpt/model/base.py [hunk 15]
```diff
@@ -124,6 +132,8 @@ class GPTConfig(TransformerConfig, io.IOMixin):
     seq_length: int = 1024
     attention_softmax_in_fp32: bool = False
     masked_softmax_fusion: bool = True
+    cross_entropy_loss_fusion: bool = True
```
Comment (gautham-kollu) [inline nemo/collections/llm/gpt/model/base.py:135]: I am unable to see the usage of this ?

File: nemo/collections/llm/recipes/llama3_70b.py [hunk 16]
```diff
@@ -93,6 +95,26 @@ def pretrain_recipe(
...
+    name: str, ckpt_dir: str, num_nodes: int, num_gpus_per_node: int, fn: Callable = pretrain
+) -> Partial:
```
Comment (gautham-kollu) [inline nemo/collections/llm/recipes/llama3_70b.py:100]: Doc-string that explains this is a performant version and is prone to some errors. 

link the fallback to pretrain_recipe() for safer.

File: nemo/collections/llm/recipes/llama3_70b.py [hunk 17]
```diff
@@ -93,6 +95,26 @@ def pretrain_recipe(
...
+    recipe = pretrain_recipe(
+        name=name, ckpt_dir=ckpt_dir, num_nodes=num_nodes, num_gpus_per_node=num_gpus_per_node, fn=fn
+    )
+
+    recipe.trainer.callbacks.append(
+        Config(
+            MegatronCommOverlapCallback,
+            tp_comm_overlap=True,
+            tp_comm_overlap_cfg=userbuffers_bf16_h100_h8192_tp4_mbs1_seqlen8192,
```
Comment (gautham-kollu) [inline nemo/collections/llm/recipes/llama3_70b.py:109]: Can we wrap this in a getter method based on 
 - precision, hardware , seqlen, batch-size, TP etc
 
np: (optional) runtime detect the hardware or please add a TODO.

File: nemo/collections/llm/recipes/llama3_8b.py [hunk 18]
```diff
@@ -92,6 +93,22 @@ def pretrain_recipe(
...
+    name: str, ckpt_dir: str, num_nodes: int, num_gpus_per_node: int, fn: Callable = pretrain
```
Comment (gautham-kollu) [inline nemo/collections/llm/recipes/llama3_8b.py:97]: Missing DocString

File: nemo/collections/llm/recipes/llama3_8b.py [hunk 19]
```diff
@@ -92,6 +93,22 @@ def pretrain_recipe(
...
+) -> Partial:
+    recipe = pretrain_recipe(
+        name=name, ckpt_dir=ckpt_dir, num_nodes=num_nodes, num_gpus_per_node=num_gpus_per_node, fn=fn
+    )
+
+    recipe.trainer.callbacks.append(
+        Config(
+            MegatronCommOverlapCallback,
+            tp_comm_overlap=False,
```
Comment (gautham-kollu) [inline nemo/collections/llm/recipes/llama3_8b.py:106]: why is it false here ?

File: nemo/lightning/pytorch/callbacks/megatron_comm_overlap.py [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+    align_param_gather: bool = None
+    bucket_size: int = None
+    # Pipeline bubble overlap
+    defer_embedding_wgrad_compute: bool = None
+    wgrad_deferral_limit: int = None
+
+
+class MegatronCommOverlapCallback(Callback):
+    """
+    A PyTorch Lightning callback to enable communication compute overlap.
```
Comment (gautham-kollu) [inline nemo/lightning/pytorch/callbacks/megatron_comm_overlap.py:48]: 💯 on the docstring

**Comment by gautham-kollu** [comment]
LGTM! Besides a few docstrings and getter apis. Approving.

**Review by gautham-kollu — APPROVED** [review]
LGTM! Besides a few docstrings and getter apis. Approving.

File: nemo/collections/llm/gpt/model/base.py (hunk 15 above)
Comment (JimmyZhang12) [reply nemo/collections/llm/gpt/model/base.py:135]: When we compute the loss this enables the use of a fused kernel:
https://github.com/NVIDIA/Megatron-LM/blob/5ec1e292822a5df9206bc04d371497b5d0e4ed4c/megatron/core/models/common/language_module/language_module.py#L38
While has better perf and uses less memory

File: nemo/collections/llm/recipes/llama3_70b.py (hunk 17 above)
Comment (JimmyZhang12) [reply nemo/collections/llm/recipes/llama3_70b.py:109]: Something like:
`get_tp_comm_overlap_config(precision, hardware, seqlen, etc) ?

File: nemo/collections/llm/recipes/llama3_8b.py (hunk 19 above)
Comment (JimmyZhang12) [reply nemo/collections/llm/recipes/llama3_8b.py:106]: See https://github.com/NVIDIA/NeMo/pull/10209#discussion_r1745017511
Maybe Im misunderstanding this?

**Review by akoumpa — APPROVED** [review]
